### DISCUSSION TITLE
Switch between Multiple Databases in DI

### DISCUSSION THREAD

**Discussion by bobskigit** [ravendb/ravendb#13059] [Q&A] [answered]
Our API needs simultaneous access to two RavenDB databases in the same cluster, we use DI to give each service a async session.

Can we dynamically switch between each database or do we need two sessions once for each database (assuming same docstore) .

What is the best way to handle this situation?

**Comment by ayende** [comment]
In your DI, `store.OpenAsyncSession(databaseName);`  to get the session to the right database.

**Comment by ppekrol** [comment, marked as answer]
Each session can communicate only to 1 database. You would have to open 2 separate sessions. This is because each session on SaveChanges creates an ACID transaction in a target database and what you would want would be a DTC transaction (those are not supported) among multiple databases.

**Comment by bobskigit** [comment]
So I need two independent sessions when I instantiate the docstore and pass either or both session instances to the relevant services? 

One service might require access to both databases or just one of them

**Reply by ppekrol** [reply]
Opening a session is a lightweight operation. You can open both of them on each request (and dispose after) or maybe pass a document store to those services and open them inside the service if needed?

**Reply by bobskigit** [reply]
That works nicely. Longer-term we probably need to have two sessions and then the developer can choose which ones are needed in the service implementation.

Thanks for the help

**Reply by ayende** [reply]
To expand on this, the session is just some memory, not a TCP connection, etc.
The two sessions to different dbs will actually share the same TCP connection pool, etc.
